numba/tests/test_struct_ref.py

```diff
@@ -0,0 +1,273 @@
+"""
+Test mutable struct, aka, structref
+"""
+import warnings
+
+import numpy as np
+
+from numba import typed, njit, errors
+from numba.core import types, structref
+from numba.extending import overload_method
+from numba.tests.support import (
+    MemoryLeakMixin, TestCase, temp_directory, override_config,
+)
+
+
+@structref.register
+class MySimplerStructType(types.StructRef):
+    """
+    Test associated to this type represent the lowest level uses of structref.
+    """
+    pass
+
+
+my_struct_ty = MySimplerStructType(
+    fields=[("values", types.intp[:]), ("counter", types.intp)]
+)
+
+structref.define_boxing(MySimplerStructType, structref.StructRefProxy)
+
+
+class MyStruct(structref.StructRefProxy):
+
+    def __new__(cls, values, counter):
+        # Define this method to customize the constructor.
+        # The default takes `*args`. Customizing allow the use of keyword-arg.
+        # The impl of the method calls `StructRefProxy.__new__`
+        return structref.StructRefProxy.__new__(cls, values, counter)
+
+    # The below defines wrappers for attributes and methods manually
+
+    @property
+    def values(self):
+        return get_values(self)
+
+    @values.setter
+    def values(self, val):
+        return set_values(self, val)
+
+    @property
+    def counter(self):
+        return get_counter(self)
+
+    def testme(self, arg):
+        return self.values * arg + self.counter
+
+
+@structref.register
+class MyStructType(types.StructRef):
+    """Test associated with this type represent the higher-level uses of
+    structef.
+    """
+    pass
+
+
+# Call to define_proxy is needed to register the use of `MyStruct` as a
+# PyObject proxy for creating a Numba-allocated structref.
+# The `MyStruct` class and then be used in both jit-code and interpreted-code.
+structref.define_proxy(
+    MyStruct,
+    MyStructType,
+    ['values', 'counter'],
+)
```

Comment:
Is there benefit in making this into a factory? Also wonder if the getters and setters like those here https://github.com/numba/numba/pull/5978/files#diff-9b2fbc1beb945aef47f6c71220bcd54dR101-R113 could be autogenerated in such a case, perhaps into a base class from which the other's can inherit and later be proxied? Perhaps not for now :)!

Reply:
I was trying to autogenerate the getters/setters via a base class. Just too much work for now.